pipeline/04-interpret.R

```diff
@@ -105,6 +112,115 @@ lightgbm::lgb.importance(lgbm_final_full_fit$fit) %>%
   rename_with(~ paste0(.x, "_value"), gain:frequency) %>%
   write_parquet(paths$output$feature_importance$local)
 
+
+
+
+#- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# 4. Calculate comps -----------------------------------------------------------
+#- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+if (comp_enable) {
+  message("Calculating comps")
+
+  # Calculate the leaf node assignments for every predicted value.
+  # Due to integer overflow problems with leaf node assignment, we need to
+  # chunk our data such that they are strictly less than the limit of 1073742
+  # rows. More detail here: https://github.com/microsoft/LightGBM/issues/1884
+  chunk_size <- 500000
+  chunks <- split(
+    assessment_data_prepped,
+    ceiling(seq_along(assessment_data_prepped[[1]]) / chunk_size)
+  )
+  chunked_leaf_nodes <- chunks %>%
+    map(\(chunk) {
+      predict(
+        object = lgbm_final_full_fit$fit,
+        newdata = as.matrix(chunk),
+        type = "leaf",
+      )
+    })
+  # Prefer do.call(rbind, ...) over bind_rows() because the chunks are
+  # not guaranteed to have the same number of rows, and bind_rows() will raise
+  # an error in that case
+  leaf_nodes <- do.call(rbind, chunked_leaf_nodes) %>% as_tibble()
```

Comment:
I'm not sure if this is the right approach, or if I just don't understand R very well 😅